Added Johnson-Cook Plasticity Model

TheGreatCid opened the pull request:
Added JC plasticity model for use in a variational framework. 
Additionally, return mapping is now floored at zero to avoid negative plastic strains.

Need a fix to JC documentation regarding having both a super and subscript on a single variable. 

Fixes issue #130

hugary1995 commented on src/materials/large_deformation_models/LargeDeformationJ2Plasticity.C:
```diff
@@ -37,11 +40,13 @@ LargeDeformationJ2Plasticity::updateState(ADRankTwoTensor & stress, ADRankTwoTen
   stress_dev_norm = std::sqrt(1.5 * stress_dev_norm);
   _Np[_qp] = 1.5 * stress_dev / stress_dev_norm;
-
+  _trial[_qp] = stress_dev_norm;
```
Is this _trial used anywhere?

TheGreatCid replied:
Ah no. It was a debugging parameter that I forgot to remove.

hugary1995 commented on src/materials/large_deformation_models/LargeDeformationJ2Plasticity.C:
```diff
@@ -37,11 +40,13 @@ LargeDeformationJ2Plasticity::updateState(ADRankTwoTensor & stress, ADRankTwoTen
   // Return mapping
   ADReal phi = computeResidual(stress_dev_norm, delta_ep);
   if (phi > 0)
+  {
     returnMappingSolve(stress_dev_norm, delta_ep, _console);
+  }
```
The code style here is to remove the enclosing braces if the subroutine is a one-liner.

hugary1995 commented on src/materials/large_deformation_models/LargeDeformationJ2Plasticity.C:
```diff
@@ -51,27 +56,37 @@ LargeDeformationJ2Plasticity::updateState(ADRankTwoTensor & stress, ADRankTwoTen
 
-  // Compute generated heat
-  _heat[_qp] = _hardening_model->plasticDissipation(delta_ep, _ep[_qp], 1) * delta_ep / _dt;
+  // Avoid div/0 issues
+  if (delta_ep > 0 && _ep[_qp] > 0)
+  {
+    _heat[_qp] = _hardening_model->plasticDissipation(delta_ep, _ep[_qp], 1) * delta_ep / _dt;
+  }
+  else
+  {
+    _heat[_qp] = 0;
+  }
```
I wonder if this ever happens as you've set the minimum permissible value to 0. I think either this check is redundant, or the minimum permissible value isn't doing its job. But if you want to be extra sure you can add an assert here, something like
```cpp
mooseAssert(_ep[_qp] >= 0 && delta_ep >= 0, "Error message otherwise");
```

hugary1995 replied:
> Issues arrive at ep=0 for the log term in the plastic dissipation of the JC model. Nothing to do with the return mapping. Since `_heat[qp]` is normally called regardless of the existence of plastic strain, it is necessary to add a conditional to avoid the log(0) issue. Let me know if there is a more elegant solution I am not seeing.

Well, then you should guard against zero ep and delta_ep in `JohnsonCookHardening::plasticDissipation`, not here.

hugary1995 replied:
This is my last complaint. PR looks good otherwise.

TheGreatCid replied:
Good point. I'll add that in

hugary1995 reviewed (changes requested):
Thanks. I'll take a closer look when tests pass.

hugary1995 commented:
The precheck is failing because your commit messages don't contain a ticket number. Just add "close #130" to any commit message.

hugary1995 commented:
A regression test is diffing.

TheGreatCid commented:
Interesting. I'll look into that

TheGreatCid commented:
Fixed the issue. 
I introduced a new material property for heat that was clashing with what already existed. 
I removed that, and now everything is working properly.

TheGreatCid commented:
You were right, I did not need that floor protection to begin with
